doc/build/sysbuild/index.rst

```diff
@@ -648,3 +693,285 @@ each image that is part of a project. Alternatively, there are
+=========================================
+
+Just like for individual Zephyr applications, you can add configuration or
+flashing dependencies for groups of these applications.
+See :ref:`sysbuild_zephyr_application_dependencies` for more information.
+
+Here is an example of group-wise dependencies:
+
+.. code-block:: cmake
+
+   sysbuild_add_dependencies(GROUP CONFIGURE my_group group_a group_b)
+   sysbuild_add_dependencies(GROUP FLASH     my_group group_a group_b)
+
+This will ensure that every image in ``group_a`` and ``group_b`` will be
+configured and flashed before every image in ``my_group``. The ordering within
+the groups themselves is not determined here. Moreover, groups are not treated
+as units; once other image-wise and group-wise dependencies are factored in,
+the images belonging to one group are not expected to stick together in the
+sorted order.
+
+Note that ``my_group`` is allowed to overlap with ``group_a`` or ``group_b``.
+If two groups overlap and one group depends on the other, then the images that
+the two groups have in common will be placed in the middle of the sorted order,
+relative to the images unique to either group.
+
+To help illustrate what this means, consider a different example:
+
+.. code-block:: cmake
+
+   ExternalZephyrProject_Group(
+     group_a
+     INCLUDE i_1 i_2 i_3 i_4 i_5 i_6
+   )
+   ExternalZephyrProject_Group(
+     group_b
+     INCLUDE i_1 i_2 i_3 i_7 i_8 i_9
+   )
+   sysbuild_add_dependencies(GROUP FLASH group_a group_b)
+
+The flashing order represented by this dependency would be something like this:
+
+.. code-block:: none
+
+      ┌───────────────────────────── unique to group_b
+      │                     ┌─────── unique to group_a
+     i_7        i_1        i_4
+   { i_8 } -> { i_2 } -> { i_5 }
+     i_9        i_3        i_6
+                 └────────────────── intersection
+
+In terms of image-wise sysbuild dependencies, this is equivalent to:
+
+.. code-block:: cmake
+
+   sysbuild_add_dependencies(IMAGE FLASH i_1 i_7 i_8 i_9)
+   sysbuild_add_dependencies(IMAGE FLASH i_2 i_7 i_8 i_9)
+   sysbuild_add_dependencies(IMAGE FLASH i_3 i_7 i_8 i_9)
+
+   sysbuild_add_dependencies(IMAGE FLASH i_4 i_1 i_2 i_3 i_7 i_8 i_9)
+   sysbuild_add_dependencies(IMAGE FLASH i_5 i_1 i_2 i_3 i_7 i_8 i_9)
+   sysbuild_add_dependencies(IMAGE FLASH i_6 i_1 i_2 i_3 i_7 i_8 i_9)
+
+If you are not sure how this works when a group has multiple dependencies,
+remember that the ``sysbuild_add_dependencies()`` function is associative,
+so the following is true:
+
+.. code-block:: cmake
+
+   sysbuild_add_dependencies(GROUP CONFIGURE g_1 g_2 g_3 ... g_n)
+
+   # This is equivalent to:
+   sysbuild_add_dependencies(GROUP CONFIGURE g_1 g_2)
+   sysbuild_add_dependencies(GROUP CONFIGURE g_1 g_3)
+   ...
+   sysbuild_add_dependencies(GROUP CONFIGURE g_1 g_n)
+
+.. tip::
+
+   You are allowed to use ``sysbuild_add_dependencies()`` with non-existent
+   groups, just as well as with non-existent images.
```

Comment:
And why?

Reply:
> For groups, I still don't want to raise an error. 

Raising errors when something is not defined is generally a good things.
It catches a lot of small mistakes early and avoid lot of wasted debugging time when something doesn't work as expected later during build or flash.

For example:
```
sysbuild_add_dependencies(GROUP CONFIGURE botloader ...) # <---- whoops, should have been bootloader, but not discovered cause non-existing groups are allowed.
```

Reply:
I understand, but raising an error would also clash with images/groups and their dependencies being defined out of order. See: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1193611885

Reply:
> but raising an error would also clash with images/groups and their dependencies being defined out of order.

dependencies should be defined after images / (groups) has been defined.
So that adding dependencies on something not defined fails immediately where the `sysbuild_add_dependencies()` is called. Anything else will make it much harder to debug, either because they are silently ignored, or because it's being warned / error at completely different location or at build time.

Such errors causes confusion as users must suddenly try to search for the location where the dependency is defined.

Reply:
> dependencies should be defined after images / (groups) has been defined.

Okay, then one option is to allow adding dependencies in `PRE_CMAKE`, as you suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1209993063

> So that adding dependencies on something not defined fails immediately

I think people might expect this function to behave the same way as the standard `add_dependencies(<target> <dependencies>)`. It only fails immediately if `<target>` doesn't exist, while `<dependencies>` are not checked until the end of the configuration stage.

So, that's what I would probably add for sysbuild, with the final checks being deferred to `sysbuild_images_sort()`.

Reply:
> adding dependencies on something not defined fails immediately where the `sysbuild_add_dependencies()` is called.

Addressed.

> a back-reference in case of a failure to where `sysbuild_add_dependencies()` was called would be great, but hard to achieve wrt. the line number.

Yeah, I wish I could add such back-references with correct filenames and line numbers. I tried out a few ways, but I couldn't get it right.